Audit log infra

vidyaap opened the pull request:
Closes #31 

Changes:
- Create a basic `LogEntry` type
- Add logs for registration and authentication flows (failures and successes)
- Add DB functions for creating new logs

picklenerd commented on dams/src/audit_log.rs:
```diff
@@ -0,0 +1,62 @@
+    Login,
+}
+
+/// Options for the outcome of a given action in a [`LogEntry`]
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Outcome {
+    Successful,
+    Failed,
+}
```
I think this is good for now but we may want to tie these to the actions in the future. I can see this type becoming very large if there are a range of different outcomes for different actions. It would be easy to make mistakes if that ends up happening.

Just wanted to make a note of this. As long as we keep a consistent format for the database representation of each entry, this should be easy to refactor down the line. For now, I think what you've got is the best way to do this.

vidyaap replied:
Yeah, I think you're right that we will eventually need to refactor the outcome-action setup, once we know what we can report back to the user.  For context, I think we still haven't decided the level of granularity we can return for the outcome of certain actions (e.g. for OPAUQE processes, I believe all types of failures need to appear the same to the user). @marsella, does that sounds right?

marsella replied:
Yeah, for authentication we only want to report a single failure type. You're both correct that we will want to reveal more information for future actions (e.g. if the policy engine rejects a signature later, we want to know which asset fiduciaries rejected it, and maybe why). But without those behaviors spec'd out, I don't know whether it would be best to add a context string to the existing `Failure` vs add more variants to express wider range of outcomes vs some other option.

marsella commented on dams-key-server/src/database/log.rs:
```diff
@@ -0,0 +1,28 @@
+    action: Action,
+    outcome: Outcome,
+) -> Result<Option<ObjectId>, Error> {
+    let collection = db.collection::<LogEntry>("log_entries");
```
question: I see we use this string in tests as well. Will it be used in other places, too? E.g. in a normal (non-test) run, where is `db.create_collection()` called? It's may be worth making it a named variable that other methods & modules can use, if that's the case.

vidyaap replied:
Yes, I meant to leave a question about this. I'd like to make some sort of config to hold all our table names so we can create them all in one shot in the tests. We shouldn't need to call `create_collection` in the actual code because Mongo will take care of creating collections that don't exist - it's just for some reason there seems to be a timing issue in the test setting.

marsella reviewed (changes requested):
This looks good! 

I think it would be easier to read if we separate the protocol logic from the log entries on the server side (in authenticate and register). This would look like making a second function that has the original body of the `run` function, then making the `run` function call that and create the appropriate log entry based on whether it succeeded or failed. 

The infrastructure looks good though, and I don't have concerns beyond readability.

vidyaap commented on dams-key-server/src/database/log.rs:
```diff
@@ -0,0 +1,28 @@
+    secret_id: Option<ObjectId>,
+    action: Action,
+    outcome: Outcome,
+) -> Result<Option<ObjectId>, Error> {
```
I suppose there isn't, I think I just wanted consistency across the `create` functions for different models. If that's not worth considering then I'm happy to switch the return type!

vidyaap commented:
Thanks for the review! I did want to ask if we should include the `FromStr` implementation that we talked about in this PR (or whatever will take care of returning a string even if we can't parse the `LogEntry` type out of whatever the server returns)?

marsella commented:
I think it's okay to not include the `FromStr` implementation we discussed. I was imagining that being useful in a context where the servers are writing logs to a file, and to retrieve them we have to read strings from the file and convert to log entries. In this version, we're storing them directly as log entries, so it doesn't seem like an immediate issue. We can revisit in a later phase when we receive relevant content from external parties (e.g. the policy engine).

vidyaap commented:
This branch was really outdated so I just rebased it and changed the logging accordingly. I think it looks more readable now, but I'd like feedback on:
- The `LogExt` trait I added based on [this post](https://stackoverflow.com/questions/63489847/how-do-i-log-to-a-file-when-an-error-occurs-while-using-the-question-mark-op). I wanted to keep the `?` structure of our helpers, but is this actually cleaner or is it more confusing?
- Is it a good idea to include the `DamsServerError` that led to the failure as context in the "failed" log entries?
- What should the actor field in the logs be? It used to be `UserId` and I changed it to be `AccountName` but will we always have access to an account name in flows besides registration and auth?
- Should I maybe move `ClientAction` to the dams crate and use that instead of the `Action` type defined in the audit_log mod of dams?

Thanks!

picklenerd commented on dams-key-server/src/command/authenticate.rs:
```diff
@@ -42,7 +48,14 @@ impl Authenticate {
     context: &Context,
-) -> Result<(ServerLoginStartResult<OpaqueCipherSuite>, UserId), DamsServerError> {
+) -> Result<
+    (
+        ServerLoginStartResult<OpaqueCipherSuite>,
+        UserId,
+        AccountName,
+    ),
```
I think this needs its own private struct.

picklenerd commented on dams-key-server/src/error.rs:
```diff
@@ -1,3 +1,12 @@
+use crate::database::log as Log;
```
```suggestion
use crate::database::log;
```

This is more of a Javascript pattern. We shouldn't use it in Rust code.

vidyaap replied:
Okay! I started this for `user/User` so there wouldn't be a name conflict with variables, and I kept the pattern for this too. Would changing it to just `log` conflict with the `log()` function though?

picklenerd replied:
I don't think it would. I believe the compiler will consider the context before reporting it as a conflict.

If we do end up with conflicts, we can do something like `use crate::database::log as logging` or just rename the module. I just want to make sure we stick to Rust's naming conventions.

picklenerd commented on dams-key-server/src/command/authenticate.rs:
```diff
@@ -26,9 +28,13 @@ impl Authenticate {
-            send_user_id(&mut channel, user_id).await?;
+            send_user_id(&mut channel, user_id)
+                .await
+                .log(&context.db, &account_name, None, Action::Authenticate)
```
Addressing your general comment, I really like this. It's a little noisy but I think the alternative is much noisier and less ergonomic. We might be able to come up with a clever way to reduce the number of arguments and then this would be perfect IMO.

picklenerd commented:
> Is it a good idea to include the DamsServerError that led to the failure as context in the "failed" log entries?

I think it depends on the purpose of these logs. If they're just used for auditing user actions, I don't think it's necessary. The error message would still be logged as part of the normal `tracing` based logging.

picklenerd commented:
> What should the actor field in the logs be? It used to be UserId and I changed it to be AccountName but will we always have access to an account name in flows besides registration and auth?

I think UserId would be better so that people's email addresses aren't floating around.

picklenerd commented:
> Should I maybe move ClientAction to the dams crate and use that instead of the Action type defined in the audit_log mod of dams?

Yeah I like this idea.

vidyaap commented:
> I think UserId would be better so that people's email addresses aren't floating around.

That makes sense, but the only thing with using UserIds is that in the registration function, we don't have the UserId until after register_finish, so I'm not sure how I'd pass it to the log function.

picklenerd commented:
> That makes sense, but the only thing with using UserIds is that in the registration function, we don't have the UserId until after register_finish, so I'm not sure how I'd pass it to the log function.

That's a good point. Maybe we could have a trait for it and implement it for both `UserId` and `AccountName` so that either one could be passed as a parameter.

marsella reviewed (commented):
Re-reviewing just to remove my "request changes" blocker.

picklenerd commented on dams-key-server/src/database/log.rs:
```diff
@@ -0,0 +1,27 @@
+/// Create a new [`LogEntry`] for the given actor, action, and outcome
+pub async fn create_log_entry(
+    db: &Database,
+    actor: &LogIdentifier,
```
```suggestion
    actor: impl Into<LogIdentifier>,
```

This would let us pass user ids and account id's directly.

picklenerd commented on dams-key-server/src/error.rs:
```diff
@@ -48,3 +53,32 @@ impl From<DamsServerError> for Status {
+    async fn log(
+        self,
+        db: &Database,
+        actor: &LogIdentifier,
```
```suggestion
        actor:  impl Into<LogIdentifier>,
```

picklenerd commented on dams/src/user.rs:
```diff
@@ -116,3 +116,20 @@ impl User {
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct LogIdentifier(String);
+
+impl From<UserId> for LogIdentifier {
+    fn from(user_id: UserId) -> Self {
+        Self(user_id.0)
+    }
+}
+
+impl From<AccountName> for LogIdentifier {
+    fn from(account_name: AccountName) -> Self {
+        Self(account_name.0)
+    }
+}
```
I know I said we shouldn't do this before, but we could also implement conversions for `&UserId` and `&AccountName` that automatically clone. This combined with the `impl Into<LogIdentifier>` changes I suggested above would let us pass the id or account name directly into the `log` function.

picklenerd reviewed (approved):
LGTM
